---
title: "Adjacency-constrained hierarchical clustering of Genomic Regions(loci)"
author: "Pierre Neuvial, Shubham Chaturvedi"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Adjacency-constrained hierarchical clustering of Genomic Regions(loci)}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

In this short vignette we demonstrate the use of `hicclust` function in the `adjclust` package.`hicclust` performs adjacency-constrained hierarchical clustering of genomic regions or loci.`hicclust` is a specialized version of `adjClustBand_heap` function designed specially to perform clustering of genomic regions or loci.This function implements the algorithm same as `adjClustBand_heap` that is described in the third chapter of [2]. It is an extension of the algorithm described in [3].  


```{r, message = FALSE}
library("adjclust")
library("HiTC")
```

## Loading genome wide contact maps 

`hic_imr90_40` is a list of objects of class `HTCexp`. Each of these objects corresponds to a Hi-C contact map between one chromosome and another. Below, `hic_imr90_40$chrXchrX` contains object `obj` which is the contact map corresponding to chromosome X vs chromosome X. The corresponding Hi-C data is stored as a Matrix::dsCMatrix in the slot named intdata.

```{r}
data("hic_imr90_40$chrXchrX", package="adjclust")
mat <- intdata(obj)
image(mat, lwd=0)
```

## Using hiciclust function for various input classes

###  Case 1: Input as HiTC::HTCexp object

```{r}
  h <- 3881
  class(obj)
  fit1 <- hicclust(obj, h)
```
    
### Case 2: Input as Matrix::dsCMatrix contact map

```{r}
  class(mat)
  fit2 <- hicclust(mat, h)
```

### Case 3: Input as text file

The input text file should have one line per pair of loci for which an interaction has been observed (in the format: locus1 <space> locus2 <space> signal).

Here, we first create a sample text file with the above format using sparse matrix `mat` created above. 

```{r}
  V3 <- mat@x
  V1 <- mat@Dimnames[[1]][mat@i+1]          #loci1names
  V2 <- rep(mat@Dimnames[[2]], diff(mat@p)) #loci2names
  
  l1 <- unique(c(as.numeric(V1), as.numeric(V2)))  
  
  l2 <- setdiff(c(72037:75918), l1) 
  V1 <- append(V1, l2)
  V2 <- append(V2, l2)
  V3 <- append(V3, rep(0, length(l2)))
  
  content <- cbind(as.numeric(V1), as.numeric(V2), as.numeric(V3))
  
  tf <- tempfile(fileext = ".txt")
  write.table(content, tf, sep = " ", col.names = FALSE, row.names = FALSE)
```

Now that we have a text file `tf`, we are ready to pass it as input to the function `hicclust`

```{r}
  fit3 <- hicclust(tf, h, sep = " ")  
```

In particular, the output contains an element named `merge` which describes the successive merges of the clustering, and an element `gains` which gives the improvement in the criterion optimized by the clustering at each successive merge.

```{r}
head(cbind(fit1$merge, fit1$gains))
```

The output can also be plotted as dendogram

For equivalent input in different classes, we get equivalent results.In this example, fit1, fit2 and fit3 are equivalent.

```{r}
all.equal(fit1$merge, fit2$merge)
all.equal(fit1$gains, fit2$gains)

all.equal(fit2$merge, fit3$merge)
all.equal(fit2$gains, fit3$gains)
```

## References

[1] Clayton, D. (2015). snpStats: SnpMatrix and XSnpMatrix classes and methods. R package version 1.20.0

[2] Dehman A. (2015). [Spatial clustering of linkage disequilibrium blocks for genome-wide association studies](https://tel.archives-ouvertes.fr/tel-01288568/). Phd Thesis, UniversitÃ© Paris Saclay.

[3] Dehman, A. Ambroise, C. and Neuvial, P. (2015). Performance of a blockwise approach in variable selection using linkage disequilibrium information. *BMC Bioinformatics* 16:148.

## Session information

```{r}
sessionInfo()
```
