<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Clustering of Hi-C contact maps • adjclust</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">adjclust</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/hicClust.html">Clustering of Hi-C contact maps</a>
    </li>
    <li>
      <a href="../articles/snpClust.html">Inferring Linkage Disequilibrium blocks from genotypes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pneuvial/adjclust">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Clustering of Hi-C contact maps</h1>
                        <h4 class="author">Shubham Chaturvedi, Pierre Neuvial, Nathalie Villa-Vialaneix</h4>
            
            <h4 class="date">2017-09-15</h4>
          </div>

    
    
<div class="contents">
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Hi-C is a sequencing-based molecular assay designed to measure intra and interchromosomal interactions between the DNA molecule. In particular, the identification of Topologically-Associated Domains (TADs), that is, regions of the genome in which physical interactions are frequent, provides insight into the three-dimensional organization of a genome [1].</p>
<p>Hi-C data are in the form of two-dimensional <em>contact maps</em>, i.e. matrices whose <span class="math inline">\(i,j\)</span> entry quantifies the intensity of the physical interaction between two genome regions <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> at the DNA level. In this vignette, we demonstrate the use of <code><a href="http://www.rdocumentation.org/packages/adjclust/topics/hicClust">adjclust::hicClust</a></code> to perform adjacency-constrained hierarchical agglomerative clustering (HAC) of Hi-C contact maps. The output of this function is a dendrogram, which can be cut to identify TADs. The algorithm used for adjacency-constrained (HAC) is described in the third chapter of [2].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"adjclust"</span>)
<span class="kw">library</span>(<span class="st">"HiTC"</span>)</code></pre></div>
</div>
<div id="loading-a-sample-hi-c-contact-map" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-a-sample-hi-c-contact-map" class="anchor"></a>Loading a sample Hi-C contact map</h1>
<p>The data set <code>hic_imr90_40_XX</code> is an object of class <code>HTCexp</code> which has been obtained from the <code>HiTC</code> package [3]. It is a contact map corresponding to chromosome X vs chromosome X.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">"hic_imr90_40_XX"</span>, <span class="dt">package=</span><span class="st">"adjclust"</span>)</code></pre></div>
<p>Now we have a look at the data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">HiTC::<span class="kw"><a href="http://www.rdocumentation.org/packages/HiTC/topics/mapC">mapC</a></span>(hic_imr90_40_XX)</code></pre></div>
<p><img src="hicClust_files/figure-html/unnamed-chunk-3-1.png" width="672"></p>
</div>
<div id="using-hicclust" class="section level1">
<h1 class="hasAnchor">
<a href="#using-hicclust" class="anchor"></a>Using <code>hicClust</code>
</h1>
<p><code>hicClust</code> operates directly on objects of class <code>HTCexp</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/hicClust.html">hicClust</a></span>(hic_imr90_40_XX)</code></pre></div>
<p>It is also possible to work on binned data. Below we choose a bin size of <span class="math inline">\(5 \times 10^5\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">binned &lt;-<span class="st"> </span>HiTC::<span class="kw"><a href="http://www.rdocumentation.org/packages/HiTC/topics/binningC">binningC</a></span>(hic_imr90_40_XX, <span class="dt">binsize=</span><span class="fl">5e5</span>)
HiTC::<span class="kw"><a href="http://www.rdocumentation.org/packages/HiTC/topics/mapC">mapC</a></span>(binned)</code></pre></div>
<p><img src="hicClust_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fitB &lt;-<span class="st"> </span><span class="kw"><a href="../reference/hicClust.html">hicClust</a></span>(binned)</code></pre></div>
<p>The output is of class <code>hclust</code>. In particular, it can be plotted as a dendrogram:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(fitB, <span class="dt">labels=</span><span class="ot">FALSE</span>, <span class="dt">hang=</span>-<span class="dv">1</span>, <span class="dt">ann=</span><span class="ot">FALSE</span>)</code></pre></div>
<p><img src="hicClust_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<p>Moreover, the output contains an element named <code>merge</code> which describes the successive merges of the clustering, and an element <code>gains</code> which gives the improvement in the criterion optimized by the clustering at each successive merge.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">cbind</span>(fitB$merge, fitB$gains))</code></pre></div>
<pre><code>##      [,1] [,2]          [,3]
## [1,] -117 -118  0.000000e+00
## [2,] -115 -116  0.000000e+00
## [3,] -114    2  0.000000e+00
## [4,]    3    1 -6.258488e-08
## [5,]   -1   -2  9.644000e+03
## [6,] -100 -101  9.915000e+03</code></pre>
</div>
<div id="other-types-of-input" class="section level1">
<h1 class="hasAnchor">
<a href="#other-types-of-input" class="anchor"></a>Other types of input</h1>
<p>Contacts maps can also be stored as objects of class <code>Matrix::dsCMatrix</code>, or as plain text files. These types of input are also accepted as first argument to <code>hicClust</code>.</p>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1] Dixon, Jesse R., <em>et al</em> (2012). “Topological domains in mammalian genomes identified by analysis of chromatin interactions.” Nature 485.7398 (2012): 376.</p>
<p>[2] Dehman A. (2015). <a href="https://tel.archives-ouvertes.fr/tel-01288568/">Spatial clustering of linkage disequilibrium blocks for genome-wide association studies</a>. Phd Thesis, Université Paris Saclay.</p>
<p>[3] Servant N., <em>et al</em> (2012). HiTC : Exploration of High-Throughput ‘C’ experiments. Bioinformatics</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">session_info</span>()</code></pre></div>
<pre><code>## Session info -------------------------------------------------------------</code></pre>
<pre><code>##  setting  value                       
##  version  R version 3.4.1 (2017-06-30)
##  system   x86_64, darwin15.6.0        
##  ui       X11                         
##  language (EN)                        
##  collate  en_US.UTF-8                 
##  tz       Europe/Paris                
##  date     2017-09-15</code></pre>
<pre><code>## Packages -----------------------------------------------------------------</code></pre>
<pre><code>##  package              * version  date       source                      
##  adjclust             * 0.4.0    2017-09-14 local (pneuvial/adjclust@NA)
##  backports              1.1.0    2017-05-22 CRAN (R 3.4.0)              
##  base                 * 3.4.1    2017-07-07 local                       
##  Biobase                2.36.2   2017-05-04 Bioconductor                
##  BiocGenerics         * 0.22.0   2017-04-25 Bioconductor                
##  BiocParallel           1.10.1   2017-05-03 Bioconductor                
##  Biostrings             2.44.2   2017-07-21 Bioconductor                
##  bitops                 1.0-6    2013-08-17 CRAN (R 3.4.0)              
##  compiler               3.4.1    2017-07-07 local                       
##  datasets             * 3.4.1    2017-07-07 local                       
##  DelayedArray           0.2.7    2017-06-03 Bioconductor                
##  devtools               1.13.3   2017-08-02 CRAN (R 3.4.1)              
##  digest                 0.6.12   2017-01-27 CRAN (R 3.4.0)              
##  evaluate               0.10.1   2017-06-24 CRAN (R 3.4.1)              
##  GenomeInfoDb         * 1.12.2   2017-06-09 Bioconductor                
##  GenomeInfoDbData       0.99.0   2017-08-03 Bioconductor                
##  GenomicAlignments      1.12.1   2017-05-12 Bioconductor                
##  GenomicRanges        * 1.28.4   2017-07-14 Bioconductor                
##  graphics             * 3.4.1    2017-07-07 local                       
##  grDevices            * 3.4.1    2017-07-07 local                       
##  grid                   3.4.1    2017-07-07 local                       
##  HiTC                 * 1.20.0   2017-04-25 Bioconductor                
##  htmltools              0.3.6    2017-04-28 CRAN (R 3.4.0)              
##  IRanges              * 2.10.2   2017-05-25 Bioconductor                
##  knitr                  1.16     2017-05-18 CRAN (R 3.4.0)              
##  lattice                0.20-35  2017-03-25 CRAN (R 3.4.1)              
##  magrittr               1.5      2014-11-22 CRAN (R 3.4.0)              
##  Matrix                 1.2-10   2017-05-03 CRAN (R 3.4.1)              
##  matrixStats            0.52.2   2017-04-14 CRAN (R 3.4.0)              
##  memoise                1.1.0    2017-04-21 CRAN (R 3.4.0)              
##  methods              * 3.4.1    2017-07-07 local                       
##  parallel             * 3.4.1    2017-07-07 local                       
##  RColorBrewer           1.1-2    2014-12-07 CRAN (R 3.4.0)              
##  Rcpp                   0.12.12  2017-07-15 CRAN (R 3.4.1)              
##  RCurl                  1.95-4.8 2016-03-01 CRAN (R 3.4.0)              
##  rmarkdown              1.6      2017-06-15 CRAN (R 3.4.0)              
##  rprojroot              1.2      2017-01-16 CRAN (R 3.4.0)              
##  Rsamtools              1.28.0   2017-04-25 Bioconductor                
##  rtracklayer            1.36.4   2017-07-13 Bioconductor                
##  S4Vectors            * 0.14.3   2017-06-03 Bioconductor                
##  snpStats               1.26.0   2017-04-25 Bioconductor                
##  splines                3.4.1    2017-07-07 local                       
##  stats                * 3.4.1    2017-07-07 local                       
##  stats4               * 3.4.1    2017-07-07 local                       
##  stringi                1.1.5    2017-04-07 CRAN (R 3.4.0)              
##  stringr                1.2.0    2017-02-18 CRAN (R 3.4.0)              
##  SummarizedExperiment   1.6.3    2017-05-29 Bioconductor                
##  survival               2.41-3   2017-04-04 CRAN (R 3.4.1)              
##  tools                  3.4.1    2017-07-07 local                       
##  utils                * 3.4.1    2017-07-07 local                       
##  withr                  2.0.0    2017-07-28 CRAN (R 3.4.1)              
##  XML                    3.98-1.9 2017-06-19 CRAN (R 3.4.1)              
##  XVector                0.16.0   2017-04-25 Bioconductor                
##  yaml                   2.1.14   2016-11-12 CRAN (R 3.4.0)              
##  zlibbioc               1.22.0   2017-04-25 Bioconductor</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#loading-a-sample-hi-c-contact-map">Loading a sample Hi-C contact map</a></li>
      <li><a href="#using-hicclust">Using <code>hicClust</code></a></li>
      <li><a href="#other-types-of-input">Other types of input</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://www.math-evry.cnrs.fr/members/cambroise/">Christophe Ambroise</a>, Shubham Chaturvedi, Alia Dehman, Michel Koskas, <a href="https://www.math.univ-toulouse.fr/~pneuvial/">Pierre Neuvial</a>, Guillem Rigaill, <a href="http://nathalievilla.org">Nathalie Villa-Vialaneix</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
