<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Inferring Linkage Disequilibrium blocks from genotypes • adjclust</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">adjclust</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/hicClust.html">Clustering of Hi-C contact maps</a>
    </li>
    <li>
      <a href="../articles/snpClust.html">Inferring Linkage Disequilibrium blocks from genotypes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pneuvial/adjclust/issues">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Inferring Linkage Disequilibrium blocks from genotypes</h1>
                        <h4 class="author">Pierre Neuvial, Shubham Chaturvedi</h4>
            
            <h4 class="date">2017-09-14</h4>
          </div>

    
    
<div class="contents">
<p>In this vignette we demonstrate the use of <code>snpClust</code> function in the <code>adjclust</code> package. <code>snpClust</code> performs adjacency-constrained hierarchical clustering of single nucleotide polymorphisms (SNPs), where the similarity between SNPs is defined by linkage disequilibrium (LD).</p>
<p>This function implements the algorithm described in the third chapter of [2]. It is an extension of the algorithm described in [3]. Denoting by <span class="math inline">\(p\)</span> the number of SNPs to cluster and assuming that the similarity between SNPs whose indices are more distant than <span class="math inline">\(h\)</span>, its time complexity is <span class="math inline">\(O(p (\log(p) + h))\)</span>, and its space complexity is <span class="math inline">\(O(hp)\)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"adjclust"</span>)</code></pre></div>
<div id="loading-and-displaying-genotype-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-and-displaying-genotype-data" class="anchor"></a>Loading and displaying genotype data</h2>
<p>The begining of this vignette closely follows the “LD vignette” of the SnpStats package [1]. First, we load genotype data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(<span class="st">"matrixStats"</span>)
<span class="kw">library</span>(<span class="st">"snpStats"</span>)
<span class="kw">data</span>(<span class="st">"ld.example"</span>, <span class="dt">package=</span><span class="st">"snpStats"</span>)</code></pre></div>
<p>We focus on the <code>ceph.1mb</code> data. These data are drawn from the International HapMap Project and concern 603 SNPs over a 1Mb region of chromosome 22 in sample of 90 Europeans.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ceph.1mb</code></pre></div>
<pre><code>## A SnpMatrix with  90 rows and  603 columns
## Row names:  NA06985 ... NA12892 
## Col names:  rs5993821 ... rs5747302</code></pre>
<p>We can compute and display the LD between these SNPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ld.ceph &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/snpStats/topics/ld">ld</a></span>(ceph.1mb, <span class="dt">stats=</span><span class="st">"R.squared"</span>, <span class="dt">depth=</span>p<span class="dv">-1</span>)
<span class="kw">image</span>(ld.ceph, <span class="dt">lwd=</span><span class="dv">0</span>)</code></pre></div>
<p><img src="snpClust_files/figure-html/unnamed-chunk-5-1.png" width="672"></p>
</div>
<div id="adjacency-constrained-hierarchical-agglomerative-clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#adjacency-constrained-hierarchical-agglomerative-clustering" class="anchor"></a>Adjacency-constrained Hierarchical Agglomerative Clustering</h2>
<p>The above figure suggests that the LD signal is concentrated close to the diagonal. Therefore, we focus on a diagonal band:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">h &lt;-<span class="st"> </span><span class="dv">100</span>
ld.ceph &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/snpStats/topics/ld">ld</a></span>(ceph.1mb, <span class="dt">stats=</span><span class="st">"R.squared"</span>, <span class="dt">depth=</span>h)
<span class="kw">image</span>(ld.ceph, <span class="dt">lwd=</span><span class="dv">0</span>)</code></pre></div>
<p><img src="snpClust_files/figure-html/unnamed-chunk-6-1.png" width="672"></p>
<p>In this unrealistically small example with only 90 subjects, it turns out that one of the LD estimates is NA due to the lack of genetic diversity in the sample:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">any</span>(<span class="kw">is.na</span>(ld.ceph))</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">which</span>(<span class="kw">is.na</span>(ld.ceph), <span class="dt">arr.ind =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>##           row col
## rs5748919 286 316</code></pre>
<p>As a quick fix for this vignette, we modify one of the observed genotypes as follows, and recalculate the LD:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ceph.1mb[<span class="dv">4</span>, <span class="dv">286</span>]@.Data[<span class="dv">1</span>,<span class="dv">1</span>] &lt;-<span class="st"> </span><span class="kw">as.raw</span>(<span class="dv">3</span>)
ld.ceph &lt;-<span class="st"> </span><span class="kw"><a href="http://www.rdocumentation.org/packages/snpStats/topics/ld">ld</a></span>(ceph.1mb, <span class="dt">stats=</span><span class="st">"R.squared"</span>, <span class="dt">depth=</span>h)</code></pre></div>
<p>Due to numerical errors in the LD estimation, some of the estimated values are slightly larger than 1, so we round the estimates to 1e-10:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">max</span>(ld.ceph<span class="dv">-1</span>)</code></pre></div>
<pre><code>## [1] 1.332268e-15</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ld.ceph &lt;-<span class="st"> </span><span class="kw">round</span>(ld.ceph, <span class="dt">digits=</span><span class="dv">10</span>)</code></pre></div>
</div>
<div id="using-snpclust-function-with-different-input-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#using-snpclust-function-with-different-input-classes" class="anchor"></a>Using snpClust function with different input classes</h2>
<div id="case-1-input-belongs-to-class-matrixdgcmatrix-generated-by-snpstatsld-function" class="section level3">
<h3 class="hasAnchor">
<a href="#case-1-input-belongs-to-class-matrixdgcmatrix-generated-by-snpstatsld-function" class="anchor"></a>Case 1: Input belongs to class Matrix::dgCMatrix generated by snpStats::ld function</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(ld.ceph)</code></pre></div>
<pre><code>## [1] "dgCMatrix"
## attr(,"package")
## [1] "Matrix"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit1 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(ld.ceph, h)</code></pre></div>
</div>
<div id="case-2-input-belongs-to-class-snpstatssnpmatrix" class="section level3">
<h3 class="hasAnchor">
<a href="#case-2-input-belongs-to-class-snpstatssnpmatrix" class="anchor"></a>Case 2: Input belongs to class snpStats::SnpMatrix</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">class</span>(ceph.1mb)</code></pre></div>
<pre><code>## [1] "SnpMatrix"
## attr(,"package")
## [1] "snpStats"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit2 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(ceph.1mb, h, <span class="st">"R.squared"</span>)</code></pre></div>
</div>
<div id="case3-input-belongs-class-basematrix" class="section level3">
<h3 class="hasAnchor">
<a href="#case3-input-belongs-class-basematrix" class="anchor"></a>Case3: Input belongs class base::matrix</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ceph.1mb &lt;-<span class="st"> </span><span class="kw">as</span>(ceph.1mb, <span class="st">"matrix"</span>)
<span class="kw">class</span>(ceph.1mb)</code></pre></div>
<pre><code>## [1] "matrix"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit3 &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(ceph.1mb, h, <span class="st">"R.squared"</span>)</code></pre></div>
</div>
</div>
<div id="output" class="section level2">
<h2 class="hasAnchor">
<a href="#output" class="anchor"></a>Output</h2>
<p>The output of the <code>snpClust</code> is of class “hclust”. In particular, it can be plotted as a dendrogram:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(fit1)</code></pre></div>
<p><img src="snpClust_files/figure-html/unnamed-chunk-13-1.png" width="672"></p>
<p>Moreover, the output contains an element named <code>merge</code> which describes the successive merges of the clustering, and an element <code>gains</code> which gives the improvement in the criterion optimized by the clustering at each successive merge.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">cbind</span>(fit1$merge, fit1$gains))</code></pre></div>
<pre><code>##      [,1] [,2] [,3]
## [1,]   -1   -2    0
## [2,] -255 -256    0
## [3,] -506 -507    0
## [4,] -492 -493    0
## [5,] -489 -490    0
## [6,] -487 -488    0</code></pre>
<p>For equivalent input in different classes, we get equivalent results.In this example, fit1, fit2 and fit3 are equivalent.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">all.equal</span>(fit1$merge, fit2$merge)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">all.equal</span>(fit1$gains, fit2$gains)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">all.equal</span>(fit2$merge, fit3$merge)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">all.equal</span>(fit2$gains, fit3$gains)</code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>[1] Clayton, D. (2015). snpStats: SnpMatrix and XSnpMatrix classes and methods. R package version 1.20.0</p>
<p>[2] Dehman A. (2015). <a href="https://tel.archives-ouvertes.fr/tel-01288568/">Spatial clustering of linkage disequilibrium blocks for genome-wide association studies</a>. Phd Thesis, Université Paris Saclay.</p>
<p>[3] Dehman, A. Ambroise, C. and Neuvial, P. (2015). Performance of a blockwise approach in variable selection using linkage disequilibrium information. <em>BMC Bioinformatics</em> 16:148.</p>
</div>
<div id="session-information" class="section level2">
<h2 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()</code></pre></div>
<pre><code>## R version 3.4.1 (2017-06-30)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: OS X El Capitan 10.11.6
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/fr_FR.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] snpStats_1.26.0    Matrix_1.2-10      survival_2.41-3   
## [4] matrixStats_0.52.2 adjclust_0.4.0    
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.12               compiler_3.4.1            
##  [3] RColorBrewer_1.1-2         GenomeInfoDb_1.12.2       
##  [5] XVector_0.16.0             bitops_1.0-6              
##  [7] tools_3.4.1                zlibbioc_1.22.0           
##  [9] digest_0.6.12              evaluate_0.10.1           
## [11] lattice_0.20-35            DelayedArray_0.2.7        
## [13] yaml_2.1.14                parallel_3.4.1            
## [15] GenomeInfoDbData_0.99.0    rtracklayer_1.36.4        
## [17] stringr_1.2.0              knitr_1.16                
## [19] Biostrings_2.44.2          S4Vectors_0.14.3          
## [21] IRanges_2.10.2             HiTC_1.20.0               
## [23] stats4_3.4.1               rprojroot_1.2             
## [25] grid_3.4.1                 Biobase_2.36.2            
## [27] XML_3.98-1.9               BiocParallel_1.10.1       
## [29] rmarkdown_1.6              magrittr_1.5              
## [31] backports_1.1.0            Rsamtools_1.28.0          
## [33] htmltools_0.3.6            BiocGenerics_0.22.0       
## [35] GenomicRanges_1.28.4       GenomicAlignments_1.12.1  
## [37] splines_3.4.1              SummarizedExperiment_1.6.3
## [39] stringi_1.1.5              RCurl_1.95-4.8</code></pre>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#loading-and-displaying-genotype-data">Loading and displaying genotype data</a></li>
      <li><a href="#adjacency-constrained-hierarchical-agglomerative-clustering">Adjacency-constrained Hierarchical Agglomerative Clustering</a></li>
      <li><a href="#using-snpclust-function-with-different-input-classes">Using snpClust function with different input classes</a></li>
      <li><a href="#output">Output</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Christophe Ambroise, Shubham Chaturvedi, Alia Dehman, Michel Koskas, Pierre Neuvial, Guillem Rigaill, Nathalie Villa-Vialaneix.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
