<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Inferring Linkage Disequilibrium blocks from genotypes • adjclust</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Inferring Linkage Disequilibrium blocks from genotypes">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">adjclust</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.5.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/hicClust.html">Clustering of Hi-C contact maps</a>
    </li>
    <li>
      <a href="../articles/notesCHAC.html">Implementation notes for the `adjclust` package</a>
    </li>
    <li>
      <a href="../articles/snpClust.html">Inferring Linkage Disequilibrium blocks from genotypes</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pneuvial/adjclust">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Inferring Linkage Disequilibrium blocks from genotypes</h1>
                        <h4 class="author">Shubham Chaturvedi, Pierre Neuvial, Nathalie Villa-Vialaneix</h4>
            
            <h4 class="date">2019-12-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/pneuvial/adjclust/blob/master/vignettes/snpClust.Rmd"><code>vignettes/snpClust.Rmd</code></a></small>
      <div class="hidden name"><code>snpClust.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># this vignette is not created if snpStats is not installed</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span>(<span class="st">"snpStats"</span>)) {</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">  knitr<span class="op">::</span>opts_chunk<span class="op">$</span><span class="kw">set</span>(<span class="dt">eval =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">}</a></code></pre></div>
<pre><code>## Loading required package: snpStats</code></pre>
<pre><code>## Loading required package: survival</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>In this vignette we demonstrate the use of <code>snpClust</code> function in the <code>adjclust</code> package. <code>snpClust</code> performs adjacency-constrained hierarchical clustering of single nucleotide polymorphisms (SNPs), where the similarity between SNPs is defined by linkage disequilibrium (LD).</p>
<p>This function implements the algorithm described in the third chapter of [2]. It is an extension of the algorithm described in [3]. Denoting by <span class="math inline">\(p\)</span> the number of SNPs to cluster and assuming that the similarity between SNPs whose indices are more distant than <span class="math inline">\(h\)</span>, its time complexity is <span class="math inline">\(O(p (\log(p) + h))\)</span>, and its space complexity is <span class="math inline">\(O(hp)\)</span>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"adjclust"</span>)</a></code></pre></div>
</div>
<div id="loading-and-displaying-genotype-data" class="section level1">
<h1 class="hasAnchor">
<a href="#loading-and-displaying-genotype-data" class="anchor"></a>Loading and displaying genotype data</h1>
<p>The begining of this vignette closely follows the “LD vignette” of the SnpStats package [1]. First, we load genotype data.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="st">"matrixStats"</span>)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"ld.example"</span>, <span class="dt">package =</span> <span class="st">"snpStats"</span>)</a></code></pre></div>
<p>We focus on the <code>ceph.1mb</code> data.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">geno &lt;-<span class="st"> </span>ceph<span class="fl">.1</span>mb[, <span class="dv">-316</span>]  <span class="co">## drop one SNP leading to one missing LD value</span></a>
<a class="sourceLine" id="cb7-2" data-line-number="2">p &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(geno)</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">nSamples &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(geno)</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">geno</a></code></pre></div>
<pre><code>## A SnpMatrix with  90 rows and  602 columns
## Row names:  NA06985 ... NA12892 
## Col names:  rs5993821 ... rs5747302</code></pre>
<p>These data are drawn from the International HapMap Project and concern 602 SNPs<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> over a 1Mb region of chromosome 22 in sample of 90 Europeans.</p>
<p>We can compute and display the LD between these SNPs.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1">ld.ceph &lt;-<span class="st"> </span>snpStats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/snpStats/man/ld.html">ld</a></span>(geno, <span class="dt">stats =</span> <span class="st">"R.squared"</span>, <span class="dt">depth =</span> p<span class="dv">-1</span>)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/graphics/image.html">image</a></span>(ld.ceph, <span class="dt">lwd =</span> <span class="dv">0</span>)</a></code></pre></div>
<p><img src="snpClust_files/figure-html/LD-1.png" width="700"></p>
</div>
<div id="adjacency-constrained-hierarchical-agglomerative-clustering" class="section level1">
<h1 class="hasAnchor">
<a href="#adjacency-constrained-hierarchical-agglomerative-clustering" class="anchor"></a>Adjacency-constrained Hierarchical Agglomerative Clustering</h1>
<p>The <code>snpClust</code> function can handle genotype data as an input:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(geno, <span class="dt">stats =</span> <span class="st">"R.squared"</span>)</a></code></pre></div>
<pre><code>## Warning in run.snpClust(x, h = h, stats = stats): Forcing the LD similarity
## to be smaller than or equal to 1</code></pre>
<pre><code>## Note: 136 merges with non increasing heights.</code></pre>
<p>Note that due to numerical errors in the LD estimation, some of the estimated LD values may be slightly larger than 1. These values are rounded to 1 internally.</p>
<p>The above figure suggests that the LD signal is concentrated close to the diagonal. We can focus on a diagonal band with the bandwidth parameter <code>h</code>:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">fitH &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(geno, <span class="dt">h =</span> <span class="dv">100</span>, <span class="dt">stats =</span> <span class="st">"R.squared"</span>)</a></code></pre></div>
<pre><code>## Warning in run.snpClust(x, h = h, stats = stats): Forcing the LD similarity
## to be smaller than or equal to 1</code></pre>
<pre><code>## Note: 134 merges with non increasing heights.</code></pre>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">fitH</a></code></pre></div>
<pre><code>## 
## Call:
## run.adjclust(mat = mat, type = type, h = h)
## 
## Cluster method   : snpClust 
## Number of objects: 602</code></pre>
</div>
<div id="output" class="section level1">
<h1 class="hasAnchor">
<a href="#output" class="anchor"></a>Output</h1>
<p>The output of the <code>snpClust</code> is of class <code>chac</code>. In particular, it can be plotted as a dendrogram silently relying on the function <code>plot.dendrogram</code>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(fitH, <span class="dt">type =</span> <span class="st">"rectangle"</span>, <span class="dt">leaflab =</span> <span class="st">"perpendicular"</span>)</a></code></pre></div>
<pre><code>## Warning in plot.chac(fitH, type = "rectangle", leaflab = "perpendicular"): 
## Detected reversals in dendrogram: mode = 'corrected', 'within-disp' or 'total-disp' might be more relevant.</code></pre>
<p><img src="snpClust_files/figure-html/dendro-1.png" width="700"></p>
<p>Moreover, the output contains an element named <code>merge</code> which describes the successive merges of the clustering, and an element <code>gains</code> which gives the improvement in the criterion optimized by the clustering at each successive merge.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/head.html">head</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(fitH<span class="op">$</span>merge, fitH<span class="op">$</span>gains))</a></code></pre></div>
<pre><code>##      [,1] [,2]
## [1,]   -1   -2
## [2,] -255 -256
## [3,] -488 -489
## [4,] -487    3
## [5,] -486    4
## [6,] -234 -235</code></pre>
</div>
<div id="other-types-of-input" class="section level1">
<h1 class="hasAnchor">
<a href="#other-types-of-input" class="anchor"></a>Other types of input</h1>
<p>In this section we show how the <code>snpClust</code> function can also be applied directly to LD values.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">h &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb22-2" data-line-number="2">ld.ceph &lt;-<span class="st"> </span>snpStats<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/snpStats/man/ld.html">ld</a></span>(geno, <span class="dt">stats =</span> <span class="st">"R.squared"</span>, <span class="dt">depth =</span> h, <span class="dt">symmetric =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/graphics/image.html">image</a></span>(ld.ceph, <span class="dt">lwd =</span> <span class="dv">0</span>)</a></code></pre></div>
<p><img src="snpClust_files/figure-html/snpClust-LD-1.png" width="700"></p>
<p>Note that we have forced the <code><a href="https://rdrr.io/pkg/snpStats/man/ld.html">snpStats::ld</a></code> function to return a symmetric matrix. We can apply <code>snpClust</code> directly to this LD matrix (of class <code>Matrix::dsCMatrix</code>):</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1">fitL &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(ld.ceph, h)</a></code></pre></div>
<pre><code>## Note: forcing the diagonal of the LD similarity matrix to be 1</code></pre>
<pre><code>## Warning in run.snpClust(x, h = h, stats = stats): Forcing the LD similarity
## to be smaller than or equal to 1</code></pre>
<pre><code>## Note: 134 merges with non increasing heights.</code></pre>
<p><code>snpClust</code> also handles inputs of class <code><a href="https://rdrr.io/r/base/matrix.html">base::matrix</a></code>:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1">gmat &lt;-<span class="st"> </span><span class="kw">as</span>(geno, <span class="st">"matrix"</span>)</a>
<a class="sourceLine" id="cb27-2" data-line-number="2">fitM &lt;-<span class="st"> </span><span class="kw"><a href="../reference/snpClust.html">snpClust</a></span>(geno, h, <span class="dt">stats =</span> <span class="st">"R.squared"</span>)</a></code></pre></div>
<pre><code>## Note: 134 merges with non increasing heights.</code></pre>
</div>
<div id="references" class="section level1">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<p>[1] Clayton D. (2015). snpStats: SnpMatrix and XSnpMatrix classes and methods. R package version 1.20.0</p>
<p>[2] Dehman A. (2015). <a href="https://tel.archives-ouvertes.fr/tel-01288568/">Spatial clustering of linkage disequilibrium blocks for genome-wide association studies</a>. Phd Thesis, Université Paris Saclay.</p>
<p>[3] Dehman A., Ambroise C., Neuvial P. (2015). Performance of a blockwise approach in variable selection using linkage disequilibrium information. <em>BMC Bioinformatics</em>, <strong>16</strong>, 148.</p>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb29-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a></code></pre></div>
<pre><code>## R version 3.6.1 (2019-07-05)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=fr_FR.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=fr_FR.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=fr_FR.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=fr_FR.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] matrixStats_0.54.0 adjclust_0.5.9     snpStats_1.34.0   
## [4] Matrix_1.2-17      survival_2.44-1.1 
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.1          rstudioapi_0.10     knitr_1.22         
##  [4] magrittr_1.5        BiocGenerics_0.30.0 zlibbioc_1.30.0    
##  [7] splines_3.6.1       MASS_7.3-51.4       lattice_0.20-38    
## [10] R6_2.4.0            rlang_0.4.0         stringr_1.4.0      
## [13] tools_3.6.1         parallel_3.6.1      grid_3.6.1         
## [16] capushe_1.1.1       xfun_0.6            htmltools_0.3.6    
## [19] yaml_2.2.0          assertthat_0.2.1    digest_0.6.18      
## [22] rprojroot_1.3-2     pkgdown_1.4.1       crayon_1.3.4       
## [25] fs_1.3.0            memoise_1.1.0       evaluate_0.13      
## [28] rmarkdown_1.12      stringi_1.4.3       compiler_3.6.1     
## [31] desc_1.2.0          backports_1.1.4</code></pre>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>We have dropped SNP rs2401075 because it produced a missing value due to the lack of genetic diversity in the considered sample.<a href="#fnref1" class="footnote-back">↩</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#loading-and-displaying-genotype-data">Loading and displaying genotype data</a></li>
      <li><a href="#adjacency-constrained-hierarchical-agglomerative-clustering">Adjacency-constrained Hierarchical Agglomerative Clustering</a></li>
      <li><a href="#output">Output</a></li>
      <li><a href="#other-types-of-input">Other types of input</a></li>
      <li><a href="#references">References</a></li>
      <li><a href="#session-information">Session information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://www.math-evry.cnrs.fr/members/cambroise/">Christophe Ambroise</a>, Shubham Chaturvedi, Alia Dehman, Michel Koskas, <a href="https://www.math.univ-toulouse.fr/~pneuvial/">Pierre Neuvial</a>, Guillem Rigaill, Nathalie Vialaneix.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
